### ISSUE TITLE
expose `manual_seed` to Hasktorch to be able to sample deterministically

### ISSUE THREAD

**Issue by tscholak** [hasktorch/hasktorch#209]
it would be beneficial to expose libtorch's `manual_seed` to Hasktorch in order to make random sampling deterministic and to avoid `IO`.

this change should affect at least the following functions and type classes:
* `rand` and `randLike`
* `randn` and `randnLike`
* `dropout`
* `Randomizable` type class
* tbc

**Comment by junjihashimoto** [comment]
@tscholak 
The manual_seed function is already implemented. (I've forgotten it.)
https://github.com/hasktorch/hasktorch/blob/master/libtorch-ffi/src/ATen/Managed/Type/Context.hs#L83-L86

**Comment by junjihashimoto** [comment]
Could you call manual_seed before using rand-function?

**Comment by junjihashimoto** [comment]
manual_seed works.

```
$ git diff
diff --git a/examples/static-mnist-mlp/Main.hs b/examples/static-mnist-mlp/Main.hs
index 955c567..e1ac9a0 100644
--- a/examples/static-mnist-mlp/Main.hs
+++ b/examples/static-mnist-mlp/Main.hs
@@ -32,6 +32,7 @@ import qualified ATen.Cast                     as ATen
 import qualified ATen.Class                    as ATen
 import qualified ATen.Type                     as ATen
 import qualified ATen.Managed.Type.Tensor      as ATen
+import qualified ATen.Managed.Type.Context     as ATen
 import           Torch.Static
 import           Torch.Static.Native     hiding ( linear )
 import           Torch.Static.Factories
@@ -161,6 +162,7 @@ main = do
       (numIters, printEvery) = (1000000, 250)
       dropoutProb            = 0.5
   (trainingData, testData) <- I.initMnist
+  ATen.manual_seed_L 123
   init                     <- A.sample
     (MLPSpec @D.Float @I.DataDim @I.ClassDim @HiddenFeatures0 @HiddenFeatures1
       dropoutProb
~/git/hasktorch/examples (feature/rand)
$ .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/static-mnist-mlp/static-mnist-mlp
Iteration: 250. Training batch loss: 16.580194. Test loss: 3.9959688. Test error-rate: 0.5345459
Iteration: 500. Training batch loss: 9.476836. Test loss: 2.2750473. Test error-rate: 0.41503906
^C
~/git/hasktorch/examples (feature/rand)
$ .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/static-mnist-mlp/static-mnist-mlp
Iteration: 250. Training batch loss: 16.580194. Test loss: 3.9959688. Test error-rate: 0.5345459
Iteration: 500. Training batch loss: 9.476836. Test loss: 2.2750473. Test error-rate: 0.41503906
^C
```

**Comment by tscholak** [comment]
Nice, good to know.
Would you say it makes sense to change, for example, dropout to:

1. Look up and cache the current random seed (is there an ATen function for getting the seed?).
2. Set the random seed via manual_seed to a value given as parameter to the dropout function.
3. Call dropout.
4. Revert the seed to the value cached in step 1.

Would that make dropout referentially transparent and the effects to libtorch’s rng unobservable? If so, then we could get rid of IO in the return type.

**Comment by austinvhuang** [comment]
Yes, as you mention the missing piece is getting the next state. I think of samplers as deterministic functions like this (similar to a state monad):

seed :: Int -> generator
sample :: generator -> (value, generator)

One point of comparison - 0.1 implementation had an explicit notion of a generator object:

https://github.com/hasktorch/hasktorch/blob/dev/hasktorch/utils/Torch/Core/Random.hs

In that case the internal state of the generator was tracked in IO (we didn't get around to looking into if the return value approach above was feasible). Can also see how other libraries do it:

https://github.com/patperry/hs-gsl-random/blob/master/lib/GSL/Random/Dist.hs
https://github.com/amcphail/hmatrix-gsl-stats/blob/master/lib/Numeric/GSL/Distribution/Continuous.hs

One potential limitation in implementing this mechanism though - at a quick glance the rng object doesn't seem to be surfaced / returned in the libtorch api. Will have a closer look later.

The Jax design has some nice properties - worth seeing if it's feasible to replicate those: https://github.com/google/jax/blob/master/design_notes/prng.md

**Comment by tscholak** [comment]
@austinvhuang ah, yes, the state monad approach :)

ok, then how about this:

```
dropout :: RNG -> Tensor dtype shape -> (RNG, Tensor dtype shape)
```
implementation:

1. cache the current state of libtorch’s internal random number generator as `s’`
2. set seed of libtorch’s internal rng to the value of the first parameter of type `RNG`
3. run dropout, assign result to `t`
4. look up libtorch’s internal rng state and assign it to `s`
5. set seed of libtorch’s rng back to `s’`
6. return `(s, t)`

thoughts?

**Comment by tscholak** [comment]
PS: incidentally, what I propose above is very close to how `Data.Random` is used, see, for instance, http://hackage.haskell.org/package/normaldistribution-1.1.0.3/docs/Data-Random-Normal.html#v:normal

**Comment by junjihashimoto** [comment]
@tscholak 
We can set custom random-generator to _fused_dropout.
_fused_dropout is a internal function of dropout.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Dropout.cpp#L85-L99

**Comment by austinvhuang** [comment]
@tscholak @junjihashimoto is it clear which libtorch API calls implement the above design? I was looking around the libtorch docs and it wasn't obvious how to get the seed corresponding to the next state. 

I kind of wish the internal representation of the RNG was exposed as a resource as it is with GSL. IIRC there were explicit RNG objects in the old C TH APIs too. I don't see a unified RNG representation in the libtorch API, but maybe I'm not looking in the right place.

As described, it feels like we're recreating it by tracking seed values on the Haskell side, but maybe that's the compromise unless the libtorch API changes to allow references to RNG objects.

@apaszke @yf225 anything we're missing in terms of the libtorch RNG design?

**Comment by apaszke** [comment]
Well there are functions like `torch.get_rng_state`, it's just that each device has a separate RNG and so managing those gets a little bit difficult. What's the issue with setting the same seed?

**Comment by junjihashimoto** [comment]
`get_rng_state` sets random-seed of global-state.
I think we get different result, when the system uses multithread, even if we use the same seed.
